File: lab05/lab05.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):

        self.val = val
        self.left = left
        self.right = right


# In[2]:


def print_tree(node, level=0):
    if node is not None:
        print("  " * level + str(node.val))
        print_tree(node.left, level + 1)
        print_tree(node.right, level + 1)


# In[3]:


def isBST(root):
    if root is None:
        return True
    if (root.left and root.left.val >= root.val) or (root.right and root.right.val <= root.val):
        return False
    return isBST(root.left) and isBST(root.right)


# In[4]:


root = TreeNode(4)
root.left = TreeNode(3)
root.right = TreeNode(8)

root.left.left = TreeNode(1)
#root.left.right = TreeNode(1)
root.right.left = TreeNode(5)
root.right.right = TreeNode(9)

print_tree(root)


# In[6]:


print(isBST(root))


# In[2]:


get_ipython().system('jupyter nbconvert --to script lab05.ipynb')

File: lab05/lab05.java
package lab05;

public class lab05 {
    public static void printTree(TreeNode node, int level) {
        if (node != null) {
            for (int i = 0; i < level; i++) {
                System.out.print("  ");
            }
            System.out.println(node.val);
            printTree(node.left, level + 1);
            printTree(node.right, level + 1);
        }
    }
    
    public static boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }
        if ((root.left != null && root.left.val >= root.val) || 
            (root.right != null && root.right.val <= root.val)) {
            return false;
        }
        return isValidBST(root.left) && isValidBST(root.right);
    }
    
    public static void main(String[] args) {
        TreeNode root = new TreeNode(6);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);

        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(1);
        root.right.left = new TreeNode(7);
        root.right.right = new TreeNode(9);

        printTree(root, 0);
        System.out.println(isValidBST(root));
    }
}